### ISSUE TITLE
Code's equation not same with the paper

### ISSUE THREAD

**Issue by milkcoffee365** [mbrossar/ai-imu-dr#74]
In utils_numpy_filter.py.
The body velocity computation seems wrong?
![image](https://user-images.githubusercontent.com/20788393/175200498-263b8433-5f6c-438a-8d35-4a99978f2fa1.png)
the formula in paper is
![image](https://user-images.githubusercontent.com/20788393/175201539-80f78fb3-6bf9-417a-b3e2-e0c538f9a22a.png)

**Comment by scott81321** [comment]
@milkcoffee365 Was this ever sorted out?

**Comment by robocar2018** [comment]
Hello, I have some derivations.
Assume R_c_n and p_c_n are the rotation and displacement of vehicle-frame relative to imu-frame,  **which are RESOLVED in the imu-frame** at time instant n.    
Then  referenced to the world-frame, 
The IMU-frame's velocity is v_IMU_n, IMU-frame's angular velocity is R_imu_n * w_n,
Still referenced to the world-frame, assume the rigid body,
Then the vehicle-frame's velocity in the world frame would be:
v =   v_IMU_n + (R_imu_n * w_n)x r, here r  = R_imu_n * p_c_n,  which is to translate the lever arm p_c_n into world-frame

Then the vehicle-frame's velocity in the vehicle-frame would be:
v_c_n = transpose(R_c_n) * transpose(R_imu_n) * v 
           = transpose(R_c_n) * transpose(R_imu_n) *  v_IMU_n +  transpose(R_c_n) * (w_n x p)

In the above equation derivation, we use the property of   skew(R*w) = R*skew(w)*transpose(R) 

So in summary, in the equation (14), the authors missed the multiplication of matrix transpose(R_c_n)  before (w_n x p)

**Comment by vkorotkine** [comment]
Fyi, the arxiv preprint is different from the published IEEE version where this is fixed. The measurement Jacobians are also different.

**Comment by scott81321** [comment]
@vkorotkine  Are you saying their published IEEE paper has the correction mentioned by @robocar2018 ?  If so, please show us clips of the where the IEEE paper differs from the arxiv version.   I cannot access the IEEE paper, only the HAL and arxiv versions.   I would be surprised that the code would not implement exactly what is presumably in the IEEE paper including later fixes (?)

**Comment by robocar2018** [comment]
@vkorotkine , thanks for your reminder. @scott81321 ,  Vkorotkine is correct,  the IEEE version fixed the error.
Below is a link for the IEEE version, please see the equation (14), multiplication of matrix transpose(R_c_n) is applied before (w_n x p_c_n). 
https://www.researchgate.net/publication/339920913_AI-IMU_Dead-Reckoning  

Furthermore, the measurement matrix H in the IEEE version is also fixed.

**Comment by scott81321** [comment]
@vkorotkine   You're right.  I see the change in the paper in terms of bracketing.  This is cause for concern because it means ai-imu-dr code, at least for that eq. 14, does not match the paper.  Mind you, this fix does not solve my current test problem.  That 2nd term is a small contribution,  for my case anyway.   

Now, you mention changes in the Jacobians.  Please tell us what they are!

**Comment by robocar2018** [comment]
@scott81321 ,  here is the result of my derivation for the H
H  =  A[0, R_IMU_T_n, 0, skew(p_c_n),  0, B, C], with A the same as paper's definition.
B = skew( R_IMU_T_n * v_IMU_n +  (w_imu_n - b_w_n) x p_c_n )
C = skew(w_imu_n - b_w_n)
Basically, B is different from IEEE paper's. 
And both skew(p_c_n) and C have a sign difference.

**Comment by scott81321** [comment]
@robocar2018 OMG so even the IEEE paper has an error?  Any other discrepancies you noticed? @vkorotkine mentioned Jacobians i.e. in the plural.

**Comment by robocar2018** [comment]
@scott81321 , for the IEEE paper, at least we can see the common extrinsic matrix is extracted out for the H matrix,  which is reflected inside the matrix A. While for the arxiv preprint, we could see the extrinsic matrix is missing in the matrix A. 
As for the matrix B, IEEE paper either has an error or a typo, since the following expression does not make sense. 
   
![image](https://user-images.githubusercontent.com/30919027/219553596-291a7aaa-6d62-472f-b1b9-2af7c07635c6.png)

**Comment by scott81321** [comment]
@robocar2018 Thank you for all this but something is still amiss.  I gather b_w_n must be 'b_omega' in the code, right?  If so, then  why does not b_omega not appear even in the fixed eq. 14 of the IEEE paper? because it certainly appears in the code!   'C' is 'Omega' in the code 't_c_i' is 'p_n_c' as you call it.  In the code, 'w' is u[:3] and you always find b_omega subtracted from it  i.e. u[:3] - b_omega.   So eq. 14 in IEEE paper would still be wrong?

**Comment by robocar2018** [comment]
@scott81321 , yes you are right.
I call it b_w_n and p_n_c based on the equation in the paper. In the code, b_w_n  is represented by b_omega (which is the estimated gyro bias), and p_n_c is represented by  t_c_i (which is lever arm between car and imu resolved in the imu-frame).  And your updated equation is correct. i.e., 
  `H_v_imu += self.skew(Omega.dot(t_c_i))`

**Comment by robocar2018** [comment]
@scott81321 , below is my updated measurement Jacobian matrix H in the code. Please let me know if you have any questions, thanks.
Basically the previous code, I found the authors forgot the common extrinsic matrix in the H. 



`def update(self, Rot, v, p, b_omega, b_acc, Rot_c_i, t_c_i, P, u, i, measurement_cov):
        # orientation of body frame
        Rot_body = Rot.dot(Rot_c_i)
        # velocity in imu frame
        v_imu = Rot.T.dot(v)
        # velocity in body frame in the imu axis
        v_body = Rot_c_i.T.dot(v_imu)
        # velocity in body frame in the vehicle axis
        v_body += self.skew(t_c_i).dot(u[:3] - b_omega)  # transalate the imu-frame-velocity to vehicle-frame-velocity
        Omega = self.skew(u[:3] - b_omega)  # skew of angular velocity
        # Jacobian w.r.t. car frame
        H_v_imu = self.skew(v_imu + Omega.dot(t_c_i)) 
        H_t_c_i = self.skew(t_c_i)
        HH = np.zeros((3, self.P_dim))  # H is a 3x21 matrix
        HH[:, 3:6] = Rot_body.T  
        HH[:, 9:12] = Rot_c_i.T.dot(H_t_c_i)
        HH[:, 15:18] = Rot_c_i.T.dot(H_v_imu)  # Jacobian of delta_imu_car_rotation_extrinsic 
        HH[:, 18:21] = Rot_c_i.T.dot(Omega)    # Jacobian of delta__imu_car_translation_extrinsic
        H = HH[1:] # extract the second and third rows from HH, which correspond to lateral and vertical speed components respectively.
        r = - v_body[1:]   # r is the residual between measurement, which is just difference between a 2x1 zero vector and v_body[1:],  
                           # v_body[1] is the lateral speed, v_body[2] is the upward speed
        R = np.diag(measurement_cov)
        # Update the state and its covariance matrix 
        Rot_up, v_up, p_up, b_omega_up, b_acc_up, Rot_c_i_up, t_c_i_up, P_up = \
            self.state_and_cov_update(Rot, v, p, b_omega, b_acc, Rot_c_i, t_c_i, P, H, r, R)
        return Rot_up, v_up, p_up, b_omega_up, b_acc_up, Rot_c_i_up, t_c_i_up, P_up
`

**Comment by scott81321** [comment]
@robocar2018  Thank you for all this.  I have implemented this code and found that it did not fix the issue I faced concerning eq. 6.  These extra terms in  'w' which involve the gyroscope readings are very very small.  In my case, I have a gyro_z lasting for a while with a vehicle experiencing changing velocities.  The algo fails here even with these corrections to the code.   The Rotation matrix Rimu of eq. 6 degrades and when the vehicle comes to a dead stop, there is no gyro 'juice' to kickstart RImu out of its bad state of the rotation matrix creates spurious accelerations.  The code has a 'normalize_rot' routine to deal with numerical error on the Rimu but it is not enough.   Any suggestions?

**Comment by robocar2018** [comment]
@scott81321, which equation are you talking about?  Equation (6) in the paper does not extra terms 'w'
![image](https://user-images.githubusercontent.com/30919027/219887948-5778645e-8682-4255-b2a4-9231dfe7ef93.png)

**Comment by scott81321** [comment]
Thanks for telling me.   You're right.  But here is an idea.  If the vehicle is at a stop,  what if one measured the mean gyro and mean accel for say a period of time, say a few seconds? During a full stop, I expect the noise to be Gaussian.  The mean if accel_x,y and gyro_x,y,z should be zero (drift not withstanding) and accel_z should be registering gravity (using KITTI format with x as forward, y is left and z is up).    On  a flat surface pitch and roll are zero and I gather so would yaw during a full stop (?)

**Comment by robocar2018** [comment]
Yes, Looks like it can work.  But I guess yaw would still drift in a longer term when the vehicle stops for a long time.  Anyway, adding ZUPT feature would definitely improve this dead reckoning performance which lacks in the current open source code.

**Comment by scott81321** [comment]
@robocar2018  Hello,  in regards to your first question, if you mean committing code to ai-imu-dr, only Martin Brossard can commit and his own commitments in industry. As for my own code, sorry but that is propriety though I can mention ideas.

**Comment by robocar2018** [comment]
No problem,  thanks.

**Comment by scott81321** [comment]
@scott81321 THX. It makes sense.  Please let us know if you find any other errors or discrepancies in the equations of the paper.  In the meantime, my kludge of estimating if the velocity is below a threshold, say 0.15 m/s, based entirely on IMU accel data works surprisingly well.  Thanks for the ZUPT heads up. Mind you, I still would like to resolve the degradation of the Rotation matrix before the algo reaches that critical point where the vehicle has stopped.  When this happens, the angle between R*g and g seems to get worse.  My yaw descends to about one but then remains level when it should gradually go to zero.    I saw your recent issue.  I am a bit puzzled.  Surely the effect of the earth's radius is not felt over a range of a few meters?

**Comment by scott81321** [comment]
Hello @robocar2018   Since you have gone to this trouble to re-derive the equations of Brossard's paper,
could you please make them public?  E.g. could you scan them and post them somehow?  

A few questions to you (or anyone else who can answer), if I may:

1) In the paper, below eq. 11, it is said that "...the covariance matrix is updated through
P_n+1 = Fn*Pn*Fn^T + Gn*Qn*Gn^T , (11) where Fn, Gn are Jacobian matrices of f(·) with respect to x_n
and u_n".  Some have read this and concluded that Fn is df/dx_n and Gn is df/du_n.     I, for one, doubt this interpretation but can you say anything about this?  

2) Do you have any info or knowledge about Gn?  It seems to be rather unique in that I don't find it explicitly for the standard Kalman filter or even the extended version.
